File: 198. House Robber/198.go
package leetcode

func rob(nums []int) int {
	if len(nums) == 0{
		return 0
	}
	first := 0
	second := nums[0]
	for i:=1; i<len(nums); i++{
		if first + nums[i] > second{
			first, second = second, first+nums[i] 
		}else{
			first = second
		}
	}
	return second
}

File: 198. House Robber/_198.py
class Solution:
    def rob(self, nums: list) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        first = nums[0]
        second = max(first, nums[1])

        for i in range(2, len(nums)):
            if  nums[i]+first>second:
                first, second = second, nums[i]+first
            else:
                first = second
        
        return second

    # 动态规划的思想
    # 这个写的简洁很多，少了判断
    # 在写go的时候，有反应减少判断，但依旧没有这个减少的好
    def rob1(self, nums):
            last, now = 0, 0
            for i in nums: last, now = now, max(last + i, now)
            return now